File: server/CustomerServer/src/main/java/CustomerServer/controller/DonHangController.java
package CustomerServer.controller;

import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import org.springframework.http.ResponseEntity;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import CustomerServer.dto.ResponseFormat;
import CustomerServer.dto.donhang.LoaiHinhThanhToan;
import CustomerServer.dto.donhang.SanPhamDonHang;
import CustomerServer.dto.donhang.TaoDonHangRequest;
import CustomerServer.dto.donhang.TrangThaiDonHang;
import lombok.AllArgsConstructor;

@RestController
@RequestMapping("/api/donhang")
@AllArgsConstructor
public class DonHangController {
  private final JdbcTemplate jdbcTemplate;

  @GetMapping("phuongthucthanhtoan")
  public ResponseEntity<?> XemDanhSachPhuongThucThanhToan() {
    return ResponseEntity.ok(ResponseFormat.success(LoaiHinhThanhToan.values()));
  }

  @GetMapping("duyetdonhang")
  public ResponseEntity<?> XemThongTinDonHangCanDuyet(String khachHangId) {
    String sql = """
        SELECT
        	sp."Id" "SanPhamId",
        	sp."Id" "PhienBanSanPhamId",
        	sp."TenSanPham",
        	sp."GiaBan",
        	sp."GianHangId",
        	sp."TenGianHang",
        	spdh."SoLuong",
        	sp."Url"
        FROM "DonHangKhachHang" dhkh
        INNER JOIN "SanPhamDonHang" spdh ON spdh."DonHangId" = dhkh."Id"
        JOIN LATERAL (
        	SELECT
        		pbsp.*,
        		sp."NguoiBanId",
        		gh."Id" "GianHangId",
        		gh."TenGianHang",
        		media."Url"
        	FROM "PhienBanSanPham" pbsp
        	INNER JOIN "SanPham" sp ON sp."Id" = pbsp."SanPhamId"
        	INNER JOIN "GianHang" gh ON gh."NguoiBanId" = sp."NguoiBanId"
        	INNER JOIN "MediaSanPham" media ON media."SanPhamId" = pbsp."SanPhamId"
        	WHERE
        		media."LoaiHinhAnhSanPham" = 0 AND
        		pbsp."Id" = spdh."PhienBanSanPhamId"
        	ORDER BY media."NgayTao" DESC
        	LIMIT 1
        ) sp ON TRUE
        WHERE
        	dhkh."KhachHangId" = ? AND
        	dhkh."LoaiHinhThanhToan" = 2
        ORDER BY dhkh."Id", spdh."SoLuong" DESC;
        """;
    List<Map<String, Object>> result = jdbcTemplate.queryForList(sql, khachHangId);
    return ResponseEntity.ok(ResponseFormat.success(result));
  }

  @GetMapping
  public ResponseEntity<?> XemDanhSachDonHang(String khachHangId) {
    String sql = """
        SELECT
          dhkh.*,
          spdh."Id" PhienBanSanPhamId,
          spdh."SoLuong",
          pbsp."TenSanPham",
          pbsp."GiaBan",
          gh."Id",
          gh."TenGianHang",
          (
        	SELECT
        	  "Url"
        	FROM "MediaSanPham" media
        	WHERE
        	  media."SanPhamId" = sp."Id"
        	  -- AND media."NgayTao" < CURRENT_DATE
        	  AND media."LoaiHinhAnhSanPham" = 0
        	ORDER BY media."NgayTao" DESC
        	LIMIT 1
          )"Url"
        FROM "DonHangKhachHang" dhkh
        LEFT JOIN "SanPhamDonHang" spdh ON spdh."DonHangId" = dhkh."Id"
        INNER JOIN "PhienBanSanPham" pbsp ON pbsp."Id" = spdh."PhienBanSanPhamId"
        INNER JOIN "SanPham" sp ON sp."Id" = pbsp."SanPhamId"
        INNER JOIN "TaiKhoanNguoiBan" nb ON nb."Id" = sp."NguoiBanId"
        INNER JOIN "GianHang" gh ON gh."NguoiBanId" = nb."Id"
        WHERE
          dhkh."KhachHangId" = ?
          AND NOT dhkh."LoaiHinhThanhToan" = 2
        ORDER BY dhkh."NgayTao" DESC, nb."Id"
        """;

    List<Map<String, Object>> data = jdbcTemplate.queryForList(sql, khachHangId);
    return ResponseEntity.ok(ResponseFormat.success(data));
  }

  @GetMapping("{donHangId}")
  public ResponseEntity<?> XemThongTinDonHang(@PathVariable String donHangId) {
    String sqlLayThongTinDonHang = """
        SELECT
        	dhkh.*,
        	tkkh."SoDienThoai",
        	dc."HoTen",
        	dc."SoDienThoai",
        	dc."DiaChiCuThe"
        FROM "DonHangKhachHang" dhkh
        INNER JOIN "TaiKhoanKhachHang" tkkh ON tkkh."Id" = dhkh."KhachHangId"
        JOIN LATERAL (
        	SELECT *
        	FROM "DiaChiGiaoHang" dcgh
        ) dc ON dc."TaiKhoanKhachHangId" = tkkh."Id"
        WHERE dhkh."Id" = ?
        ORDER BY "NgayTao" DESC
        LIMIT 1
        """;
    Map<String, Object> donHang = jdbcTemplate.queryForMap(sqlLayThongTinDonHang, donHangId);

    String sqlXemThongTinTrangThaiDonHang = """
        SELECT
        	cntt.*
        FROM "DonHangKhachHang" dhkh
        INNER JOIN "CapNhatTrangThaiDonHang" cntt ON cntt."DonHangId" = dhkh."Id"
        WHERE dhkh."Id" = ?
        ORDER BY cntt."ThoiGianTao" DESC
        """;
    List<Map<String, Object>> trangThai = jdbcTemplate.queryForList(sqlXemThongTinTrangThaiDonHang, donHangId);

    String sqlXemDanhSachSanPham = """
        SELECT
        	pbsp."SanPhamId",
        	pbsp."Id" "PhienBanId",
        	pbsp."TenSanPham",
        	pbsp."GiaBan",
        	spdh."SoLuong",
        	sp."NguoiBanId",
        	media."Url"
        FROM "DonHangKhachHang" dhkh
        INNER JOIN "SanPhamDonHang" spdh ON spdh."DonHangId" = dhkh."Id"
        INNER JOIN "PhienBanSanPham" pbsp ON pbsp."Id" = spdh."PhienBanSanPhamId"
        INNER JOIN "SanPham" sp ON sp."Id" = pbsp."SanPhamId"
        INNER JOIN "TaiKhoanNguoiBan" nb ON nb."Id" = sp."NguoiBanId"
        JOIN LATERAL (
        	SELECT "Url"
        	FROM "MediaSanPham"
        	WHERE
        		"SanPhamId" = pbsp."SanPhamId"
        		AND "LoaiHinhAnhSanPham" = 0
        	ORDER BY "NgayTao"
        	LIMIT 1
        ) media ON TRUE
        WHERE dhkh."Id" = ?
        """;
    List<Map<String, Object>> sanPham = jdbcTemplate.queryForList(sqlXemDanhSachSanPham, donHangId);
    donHang.put("trangThai", trangThai);
    donHang.put("sanPham", sanPham);

    return ResponseEntity.ok(ResponseFormat.success(donHang));
  }

  @PostMapping("duyetdonhang")
  public ResponseEntity<?> DuyetDonHang(String khachHangId) {
    List<Map<String, Object>> donHang = jdbcTemplate.queryForList("""
        SELECT
        	"Id"
        FROM "DonHangKhachHang" dhkh
        WHERE
        	dhkh."LoaiHinhThanhToan" = 2 AND
        	dhkh."KhachHangId" = ?
          """, khachHangId);

    String query = """
        UPDATE "DonHangKhachHang"
        	SET "LoaiHinhThanhToan"=?
        	WHERE "Id"=?;
          """;
    for (Map<String, Object> elem : donHang) {
      System.out.println(elem);
      jdbcTemplate.update(query, LoaiHinhThanhToan.KHI_NHAN_HANG.ordinal(), elem.get("Id").toString());
    }

    return ResponseEntity.ok(ResponseFormat.success(donHang));
  }

  @PostMapping
  public ResponseEntity<?> TaoDonHang(@RequestBody TaoDonHangRequest entity) {
    // TODO: process POST request
    // tao don hang
    // entity.setKhachHangId("khachHangId");
    var danhSachSanPham = entity.getSanPham()
        .stream()
        .filter((SanPhamDonHang i) -> i.getSoLuong() > 0)
        .toList();
    if (danhSachSanPham.isEmpty()) {
      throw new Error("yeu cau loi");
    }

    String sqlTaoDonHAng = """
        INSERT INTO "DonHangKhachHang"
        ("Id", "KhachHangId", "PhiVanChuyen", "LoaiHinhThanhToan", "NgayTao")
        VALUES (?, ?, ?, ?, ?)
        """;
    String donHangId = UUID.randomUUID().toString();
    jdbcTemplate.update(
        sqlTaoDonHAng,
        donHangId,
        entity.getKhachHangId(),
        15_000,
        entity.getLoaiHinhThanhToan().ordinal(),
        new Date());

    // cap nhat trang thai
    String sqlCapNhatTrangThaiDonHang = """
        INSERT INTO "CapNhatTrangThaiDonHang"
        ("Id", "DonHangId", "NoiDungCapNhat", "TrangThaiDonHang", "ThoiGianTao")
        VALUES (?, ?, ?, 0, CURRENT_DATE)
        """;
    String trangThaiDonHangId = UUID.randomUUID().toString();
    jdbcTemplate.update(
        sqlCapNhatTrangThaiDonHang,
        trangThaiDonHangId,
        donHangId,
        "");

    // cap nhat san pham don hang
    // var danhSachSanPham = entity.getSanPham();
    String sqlCapNhatSanPhamDonHang = """
        INSERT INTO "SanPhamDonHang"
        ("Id", "PhienBanSanPhamId", "DonHangId", "SoLuong")
        VALUES (?, ?, ? ,?)
        """;
    for (int i = 0; i < danhSachSanPham.size(); i++) {
      SanPhamDonHang item = danhSachSanPham.get(i);
      jdbcTemplate.update(
          sqlCapNhatSanPhamDonHang,
          UUID.randomUUID().toString(),
          item.getPhienBanSanPhamId(),
          donHangId,
          item.getSoLuong());
    }

    return ResponseEntity.ok(ResponseFormat.success(entity));
  }

  @DeleteMapping("duyetdonhang")
  public ResponseEntity<?> HuyDuyet(String khachHangId) {
    List<Map<String, Object>> sanPhamDonHangCanDuyet = jdbcTemplate.queryForList("""
        SELECT
        	spdh."Id" "SanPhamDonHangId",
        	dhkh."Id" "DonHangId"
        FROM "SanPhamDonHang" spdh
        INNER JOIN "DonHangKhachHang" dhkh ON dhkh."Id" = spdh."DonHangId"
        WHERE
        	dhkh."LoaiHinhThanhToan" = 2 AND
        	dhkh."KhachHangId" = ?;
          """, khachHangId);
    List<Map<String, Object>> trangThai = jdbcTemplate.queryForList("""
        SELECT
        	cn."Id" "TrangThaiId"
        FROM "CapNhatTrangThaiDonHang" cn
        INNER JOIN "DonHangKhachHang" dhkh ON dhkh."Id" = cn."DonHangId"
        WHERE
        	dhkh."LoaiHinhThanhToan" = 2 AND
        	dhkh."KhachHangId" = ?
          """, khachHangId);
    List<Map<String, Object>> donHang = jdbcTemplate.queryForList("""
        SELECT
        	"Id"
        FROM "DonHangKhachHang" dhkh
        WHERE
        	dhkh."LoaiHinhThanhToan" = 2 AND
        	dhkh."KhachHangId" = ?
                  """, khachHangId);

    for (Map<String, Object> elem : sanPhamDonHangCanDuyet) {
      String _id = (String) elem.get("SanPhamDonHangId");

      jdbcTemplate.update("DELETE FROM \"SanPhamDonHang\" WHERE \"Id\" = ?;", _id);
    }
    for (Map<String, Object> elem : trangThai) {
      String _id = (String) elem.get("TrangThaiId");
      jdbcTemplate.update("DELETE FROM \"CapNhatTrangThaiDonHang\" WHERE \"Id\" = ?;", _id);
    }
    for (Map<String, Object> elem : donHang) {
      String _id = (String) elem.get("Id");
      jdbcTemplate.update("DELETE FROM \"DonHangKhachHang\" WHERE \"Id\" = ?;", _id);
    }

    return ResponseEntity.ok(ResponseFormat.success(Map.of(
        "sanPhamDonHangCanDuyet", sanPhamDonHangCanDuyet,
        "trangThai", trangThai,
        "donHang", donHang)));
  }

  @DeleteMapping("{id}")
  public ResponseEntity<?> HuyDonHang(@PathVariable String id) {
    // TODO: process PUT request
    String sqlHuyDonHang = """
        INSERT INTO "CapNhatTrangThaiDonHang"
        	("Id", "DonHangId", "NoiDungCapNhat", "TrangThaiDonHang", "ThoiGianTao")
        VALUES (?, ?, ?, ?, CURRENT_DATE)
        """;
    jdbcTemplate.update(sqlHuyDonHang,
        UUID.randomUUID().toString(),
        id,
        "Huy Don Hang",
        TrangThaiDonHang.HUY_DON_HANG.ordinal());
    return ResponseEntity.ok(ResponseFormat.success());
  }
}

File: server/TestServer/Controllers/DonHangController.cs
using DatabaseModels;
using DatabaseModels.DTO;
using DatabaseModels.Models;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Utilities;

namespace TestServer.Controllers;

[ApiController]
[Route("api/don-hang")]
public class DonHangController(IConfiguration config, AppDbContext dbContext) : ControllerBase
{
  readonly IConfiguration _config = config;
  readonly AppDbContext dbContext = dbContext;
  readonly string serverUrl = "http://localhost:8083/api/donhang";

  [HttpPost]
  public async Task<IActionResult> TaoDonHang(int donHang = 5)
  {
    // try
    // {
    Random random = new();
    List<TaiKhoanNguoiBan> taiKhoanNguoiBan = await dbContext.TaiKhoanNguoiBan.ToListAsync();
    List<TaiKhoanKhachHang> taiKhoanKhachHang = await dbContext.TaiKhoanKhachHang.ToListAsync();
    List<SanPham> sanPham = await dbContext.SanPham.ToListAsync();
    List<GianHang> gianHang = await dbContext.GianHang.ToListAsync();
    List<PhienBanSanPham?>? phienBanSanPham = [.. sanPham
        .OrderBy(sanPham => sanPham.NguoiBanId)
        .Select(i =>
          dbContext.PhienBanSanPham
            .Where(j => j.SanPhamId == i.Id && j.NgayTao <= DateTime.UtcNow)
            .OrderByDescending(i => i.NgayTao)
            .Take(1)
            .ToList()
            .FirstOrDefault()
        )];

    List<TaoDonHangRequest> taoDonHangRequest = [];
    foreach (var kh in taiKhoanKhachHang)
    {
      for (int i = 0; i < donHang; i++)
      {
        TaiKhoanNguoiBan nguoiBan = taiKhoanNguoiBan.ElementAt(random.Next(taiKhoanNguoiBan.Count));
        List<PhienBanSanPham> _phienBanSanPham = [.. phienBanSanPham
            .Join(
              dbContext.SanPham,
              pbsp => pbsp!.SanPhamId,
              sp => sp.Id,
              (pbsp,sp)=>new {pbsp, sp})
              .Where(i=>i.sp.NguoiBanId == nguoiBan.Id)
            .Select(i=>i.pbsp!)];

        TaoDonHangRequest request = new()
        {
          KhachHangId = kh.Id,
          LoaiHinhThanhToan = random.Next(2) == 0 ? LoaiHinhThanhToanDto.VN_PAY : LoaiHinhThanhToanDto.KHI_NHAN_HANG,
          SanPham = []
        };

        int len = Math.Min(_phienBanSanPham.Count, random.Next(1, donHang));
        for (int j = 0; j < len; j++)
        {
          int _index = random.Next(_phienBanSanPham.Count);
          var item = _phienBanSanPham[_index];
          _phienBanSanPham.RemoveAt(_index);
          request.SanPham.Add(new()
          {
            PhienBanSanPhamId = item.Id,
            SoLuong = random.Next(1, 50)
          });
        }
        taoDonHangRequest.Add(request);
      }
    }

    await Parallel.ForEachAsync(taoDonHangRequest, async (data, c) =>
    {
      var body = await GenerateRequest.CreateRequest(data, serverUrl, RequestMethod.POST);
      Console.WriteLine($"TaoDonHang: {body}");
    });
    return Ok(taoDonHangRequest);
    // }
    // catch (Exception)
    // {
    //   throw;
    // }
  }

  [HttpPost("trang-thai-don-hang")]
  public async Task<IActionResult> TaoTrangThaiDonHang()
  {
    Random random = new();
    List<DonHangKhachHang> donHangKhachHang = await dbContext.DonHangKhachHang.ToListAsync();
    List<List<CapNHatTrangThaiDonHangRequest>> requests = [];
    var _trangThai = Enum.GetValues<TrangThaiDonHang>()
      .Where(i => i != TrangThaiDonHang.HUY_DON_HANG)
      .Select(i => i)
      .ToList();

    dbContext.CapNhatTrangThaiDonHang.RemoveRange(await dbContext.CapNhatTrangThaiDonHang.ToListAsync());
    await dbContext.SaveChangesAsync();
    foreach (var donHang in donHangKhachHang)
    {
      int count = random.Next(1, _trangThai.Count);
      List<CapNHatTrangThaiDonHangRequest> trangThai = [];
      for (int i = 0; i < count; i++)
      {
        trangThai.Add(new()
        {
          DonHangId = donHang.Id,
          NoiDungCapNhat = RandomGenerator.GenerateRandomString(10, 20),
          TrangThaiDonHang = _trangThai.ElementAt(i)
        });
      }
      requests.Add(trangThai);
    }
    await Parallel.ForEachAsync(requests, async (data, c) =>
    {
      foreach (var item in data)
      {
        var body = await GenerateRequest.CreateRequest(item, "http://localhost:5216/api/don-hang", RequestMethod.POST);
        Console.WriteLine($"TaoTrangThaiDonHang: {body}");
      }
    });
    return Ok(requests);
  }

  [HttpDelete]
  public async Task<IActionResult> XoaDonHang()
  {
    try
    {
      dbContext.CapNhatTrangThaiDonHang.RemoveRange(await dbContext.CapNhatTrangThaiDonHang.ToListAsync());
      dbContext.SanPhamDonHang.RemoveRange(await dbContext.SanPhamDonHang.ToListAsync());
      dbContext.DonHangKhachHang.RemoveRange(await dbContext.DonHangKhachHang.ToListAsync());
      await dbContext.SaveChangesAsync();
      return Ok();
    }
    catch (System.Exception)
    {

      throw;
    }
  }
}